### DISCUSSION TITLE
Deeply merging options when using the .set API

### DISCUSSION THREAD

**Discussion by stevenharman** [sidekiq/sidekiq#6366] [Q&A] [answered]
We can override a Job class's `sidekiq_options` on a job-by-job basis via the `.set` API: 

```ruby
class MyJob
  include Sidekiq::Job
  
  sidekiq_options queue: :within_30_seconds

  # …
end

MyJob.set(queue: :within_5_minutes).perform_async(foo, bar)
```

However, what about custom options (like for a Sidekiq extension) which have a nested data structure? For example:

```ruby
class MyJob
  include Sidekiq::Job
  
  sidekiq_options queue: :within_30_seconds, bucket: { by: :some_method, lookback: 5.minutes }

  # …
end

MyJob.set(bucket: { lookback: 1.minute }).perform_async(foo, bar)
```

Ideally, for something like this, `set(…)` would deeply-merge the `bucket:` option, rather than do a shallow merge where the entire value is replaced. So we'd end just only overriding the nested `lookback:` while preserving the class-configured `by:` option.

Is that sort of deep-merge ideal for every case? I'm not sure. I'm wondering if you (Mike) have an opinion on the matter?

**Comment by mperham** [comment]
Sidekiq does not use deep structures for this reason.

**Reply by stevenharman** [reply]
Is there an implied, "and I don't think it should, so I don't want to support a deep-merge."? 😄

**Reply by mperham** [reply, marked as answer]
If Ruby provided an easy method to do it, I would use it. But since it does not, I choose not to implement it since none of my code requires it.

**Reply by stevenharman** [reply]
That's fair. It's one of those methods from `active_support` that _seems_ like it should have made its way into stdlib by now.

I'll just work around this in my code for now. Thanks, Mike!
